### ISSUE TITLE
Cop idea: prefer &:itself then an equivalent  block

### ISSUE THREAD

**Issue by Darhazer** [rubocop/rubocop#14046]
## Is your feature request related to a problem? Please describe.

Some people will do
```ruby
collection.index_with { _1 }
```

Which I just caught by having it changed to
```ruby
collection.index_with { it }
```

For both I'd prefer 
```ruby
collection.index_with(&:itself)
```

## Describe the solution you'd like

A cop that would for blocks that contain just _1 or it, to be used as &:itself

**Comment by lovro-bikic** [comment]
I'd just add that this style:
```ruby
collection.index_with { |x| x }
```
 is also [quite common](https://github.com/search?q=%2F%5C.%5Cw%2B%20%5C%7B%20%5C%7C%5Cw%5C%7C%20%5Cw%20%5C%7D%2F%20language%3ARuby&type=code) (there are some false positives on this link), and it can also use `itself`.

**Comment by dvandersluis** [comment]
Just to note, `&:itself` is about 10% slower than any of the three styles of block (for 10k elements):

```ruby
require 'benchmark/ips'

collection = (1..10000).to_a

Benchmark.ips do |x|
  x.report('block') { collection.map { |x| x } }
  x.report('numblock') { collection.map { _1 } }
  x.report('itblock') { collection.map { it } }
  x.report('itself') { collection.map(&:itself) }
  x.compare!
end
```

```
ruby 3.4.1 (2024-12-25 revision 48d4efcb85) +PRISM [arm64-darwin24]
Warming up --------------------------------------
               block   442.000 i/100ms
            numblock   443.000 i/100ms
             itblock   437.000 i/100ms
              itself   407.000 i/100ms
Calculating -------------------------------------
               block      4.365k (± 1.9%) i/s  (229.08 μs/i) -     22.100k in   5.064494s
            numblock      4.354k (± 2.8%) i/s  (229.66 μs/i) -     22.150k in   5.091191s
             itblock      4.348k (± 3.6%) i/s  (229.98 μs/i) -     21.850k in   5.033050s
              itself      3.981k (± 1.9%) i/s  (251.18 μs/i) -     19.943k in   5.011172s

Comparison:
               block:     4365.3 i/s
            numblock:     4354.3 i/s - same-ish: difference falls within error
             itblock:     4348.1 i/s - same-ish: difference falls within error
              itself:     3981.1 i/s - 1.10x  slower
```

**Comment by bbatsov** [comment]
Good to know, although I doubt this will be a real issue for most people.

**Comment by Earlopain** [comment]
I see barely any difference on 3.4.2 between the versions. But with YJIT enabled it's actually ~3.5x slower (the block variants are about 3x faster just by enabling YJIT but `itself` gets a performance hit)

**Comment by bbatsov** [comment]
Style cops sometimes focus on readability over performance, but even if we put that aside - I imagine that whatever cop we create will be configurable, and this is just going to be one of the supported style. (I guess it can be `itself` and `it`, as I doubt anyone would go for numbered params; named params is a potential style as well)

**Comment by dvandersluis** [comment]
@bbatsov I opened #14049 to implement this cop as you suggested - configurable, with `itself` (`&:itself`), `it` (`{ it }`) and `named_parameter` (`{ |x| x }`) styles. As you suggested, I did not add a style for numblocks.
